XRPLib/imu.py

```diff
@@ -308,16 +315,22 @@ def get_pitch(self):
         :return: The pitch of the IMU in degrees
         :rtype: float
         """
-        return self.running_pitch
-    
+        if self.pitchComputed == False:
+            self.pitch = math.asin(-2 * (self.q[1]*self.q[3] - self.q[0]*self.q[2])) * 57.23
+            self.pitchComputed = True
+        return self.pitch
+
     def get_yaw(self):
         """
         Get the yaw (heading) of the IMU in degrees. Unbounded in range
 
         :return: The yaw (heading) of the IMU in degrees
         :rtype: float
         """
-        return self.running_yaw
+        if self.yawComputed == False:
+            self.yaw = math.atan2(self.q[1]*self.q[2] + self.q[0]*self.q[3], 0.5 - self.q[2]*self.q[2] - self.q[3]*self.q[3]) * 57.23
```

Comment:
The yaw reading is fine most of the time, and works while the bot is at any angle other than vertical. When the robot is rotated such that roll is near 90 (or -90), the yaw jumps up/down by a fair margin. Here's an example of it jumping from -90 to -165ish
![image](https://github.com/Open-STEM/XRP_MicroPython/assets/67293695/4e172b86-c786-4d3d-8995-8d8fb137b17f)

Reply:
> yaw jumps up/down by a fair margin

Again, this has to do with the singularity condition when near vertical. Think of a vector pointing straight out the front of the robot, projected onto the horizontal plane; the direction of that vector is what's reported as the yaw. Image the robot RPY starts at (0,0,0), then it's pitched up to 89 degrees (almost vertical). The vector pointing out from the front of the robot would get projected into the horizontal plane, and is still pointing at 0 degrees, so the yaw would still get reported as 0. Now imagine the robot is pitched 2 more degrees, going through vertical. Now projecting that vector onto the horizontal plane would cause it to point the opposite direction from the origin, so the yaw gets reported as 180 degrees (or -180).

This is all to be expected with a proper AHRS algorithm, since RPY has a singularity around the vertical directions, which is why quaternions are superior for this. Probably also worth noting that the pitch is actually bounded between [-90, 90], so when the robot gets rotated that extra 2 degrees, the pitch would still get reported as 89 degrees instead of 91 like you might expect. It's very much related to spherical coordinates.